src/main/scala/org/scanet/datasets/Dataset.scala

```diff
@@ -8,72 +8,77 @@ import scala.io.Source
 
 trait Dataset[X] {
   def iterator: Iterator[X]
-  def shape(batch: Int): Shape
 }
 
 trait Iterator[X] {
   def hasNext: Boolean
   def next(batch: Int): Tensor[X]
-  def size: Option[Int]
+  def size: Int
+  def shape(batch: Int): Shape
+}
+
+private[datasets] trait DataSource[X] {
+  def size: Int
+  def slice(from: Int, to: Int): Tensor[X]
+  def shape(batch: Int): Shape
 }
 
-case class TensorDataset[X: TensorType: Numeric](src: Tensor[X]) extends Dataset[X] {
+private[datasets] abstract class IndexedDataset[X](val ds: DataSource[X]) extends Dataset[X] {
   override def iterator: Iterator[X] = new Iterator[X] {
-    var pos: Int = 0
-    val size: Option[Int] = Some(src.shape.dims.head)
-    override def hasNext: Boolean = src.shape.dims.head > pos
-    override def next(batch: Int): Tensor[X] = {
+    private var index: Int = 0
+
+    override def hasNext: Boolean = size > index
+
+    override def next(batchSize: Int): Tensor[X] = {
       require(hasNext, "dataset has no elements left")
-      // TODO: ensure constant shape
-      val slice: Tensor[X] = src(pos until math.min(pos + batch, size.get))
-      pos = pos + slice.shape.dims.head
-      slice
+      val pos = index
+      index = index + batchSize
+      if (batchSize > size) {
+        ds.slice(pos, size) // whole source fits into batch
```

Comment:
Is that right?

Reply:
this branch handles case when we request more elements than source size, i.e. default batch size in optimizer is Int.MaxValue - so instead we just return all available rows as single (first) batch
`def shape(batch: Int)` also handles this case

Reply:
Nitpick, but you can do it more generic (which is kind of the same, but still) 
```
      if (pos + batchSize > size) {
        ds.slice(math.max(size - batchSize, 0), size)
      } else {
        ds.slice(pos, pos + batchSize)
      }
```

Reply:
Also, could you add a short path into `Tensor.slice` so in case we slice by a projection which will return the same tensor we just return `this`.